tests/Query/AggregationBuilderTest.php

```diff
@@ -0,0 +1,132 @@
+
+        // Compare with the expected pipeline
+        $expected = [
+            ['$match' => ['name' => 'John Doe']],
+            ['$limit' => 10],
+            [
+                '$addFields' => [
+                    'year' => ['$year' => ['date' => '$birthday']],
+                ],
+            ],
+            ['$sort' => ['year' => -1, 'name' => 1]],
+            ['$unset' => ['birthday']],
+        ];
+
+        $this->assertSamePipeline($expected, $pipeline->getPipeline());
+
+        // Execute the pipeline and validate the results
+        $results = $pipeline->get();
+        $this->assertInstanceOf(Collection::class, $results);
+        $this->assertCount(1, $results);
+        $this->assertInstanceOf(ObjectId::class, $results->first()['_id']);
+        $this->assertSame('John Doe', $results->first()['name']);
+        $this->assertIsInt($results->first()['year']);
+        $this->assertArrayNotHasKey('birthday', $results->first());
+
+        // Execute the pipeline and validate the results in a lazy collection
+        $results = $pipeline->cursor();
+        $this->assertInstanceOf(LazyCollection::class, $results);
+
+        // Execute the pipeline and return the first result
+        $result = $pipeline->first();
+        $this->assertIsArray($result);
+        $this->assertInstanceOf(ObjectId::class, $result['_id']);
+        $this->assertSame('John Doe', $result['name']);
+    }
+
+    public function testAddRawStage(): void
+    {
+        $collection = $this->createMock(MongoDBCollection::class);
+
+        $pipeline = new AggregationBuilder($collection);
+        $pipeline
+            ->addRawStage('$match', ['name' => 'John Doe'])
+            ->addRawStage('$limit', 10)
+            ->addRawStage('$replaceRoot', (object) ['newRoot' => '$$ROOT']);
+
+        $expected = [
+            ['$match' => ['name' => 'John Doe']],
+            ['$limit' => 10],
+            ['$replaceRoot' => ['newRoot' => '$$ROOT']],
+        ];
+
+        $this->assertSamePipeline($expected, $pipeline->getPipeline());
+    }
+
+    public function testAddRawStageInvalid(): void
+    {
+        $collection = $this->createMock(MongoDBCollection::class);
+
+        $pipeline = new AggregationBuilder($collection);
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('The stage name "match" is invalid. It must start with a "$" sign.');
+        $pipeline->addRawStage('match', ['name' => 'John Doe']);
+    }
+
+    private static function assertSamePipeline(array $expected, Pipeline $pipeline): void
+    {
+        $expected = Document::fromPHP(['pipeline' => $expected])->toCanonicalExtendedJSON();
+
+        $codec = new BuilderEncoder();
+        $actual = $codec->encode($pipeline);
+        // Normalize with BSON round-trip
+        $actual = Document::fromPHP(['pipeline' => $actual])->toCanonicalExtendedJSON();
+
+        self::assertJsonStringEqualsJsonString($expected, $actual);
```

Comment:
TIL [`assertJsonStringEqualsJsonString()`](https://docs.phpunit.de/en/9.6/assertions.html#assertjsonstringequalsjsonstring) exists. There's probably a place I could have used this in PHPLIB had I known about it.